docs: add KeptnTask ref page; enhance guide chapter for keptn-no-k8s (#2103)
keptn/lifecycle-toolkit: docs/content/en/docs/implementing/tasks-non-k8s-apps.md

title: Keptn for Non-Kubernetes Applications
description: Using Keptn with Non-Kubernetes Applications
weight: 95
---

Keptn Tasks can be triggered for workloads and applications
that are deployed outside of Kubernetes.
For example, Keptn could trigger load and performance tests

> for an application that is deployed on a virtual machine.

Author: I made it "Keptn Tasks running on a Kubernetes cluster..." and some other adjustments to make this more explicit.

To do this:

1. [Install and enable a Kubernetes cluster](#install-and-enable-a-kubernetes-cluster)
1. [Create a KeptnTaskDefinition](#create-a-keptntaskdefinition)
1. [Create and apply a KeptnTask](#create-and-apply-a-keptntask)
1. [Re-run the KeptnTask](#re-run-the-keptntask)

## Install and enable a Kubernetes cluster

You must still
[install](../install/install.md/#use-helm-chart)
Keptn on a Kubernetes cluster,
but this can be a very lightweight, single-node KinD cluster; see
[Create local Kubernetes cluster](../install/k8s.md/#create-local-kubernetes-cluster).
Keptn only triggers on-demand `KeptnTask` resources
so resource utilization is minimal.

## Create a KeptnTaskDefinition

When you have Keptn installed, create a
YAML file that defines what you want to execute

> as a `KeptnTaskDefinition` resource..

Author: fixed

title: Keptn for Non-Kubernetes Applications
description: Using Keptn with Non-Kubernetes Applications
weight: 95
---

Keptn Tasks can be triggered for workloads and applications
that are deployed outside of Kubernetes.
For example, Keptn could trigger load and performance tests
for an application that is deployed on a virtual machine.

To do this:

1. [Install and enable a Kubernetes cluster](#install-and-enable-a-kubernetes-cluster)
1. [Create a KeptnTaskDefinition](#create-a-keptntaskdefinition)
1. [Create and apply a KeptnTask](#create-and-apply-a-keptntask)
1. [Re-run the KeptnTask](#re-run-the-keptntask)

## Install and enable a Kubernetes cluster

You must still
[install](../install/install.md/#use-helm-chart)
and

> [enable](../install/install.md/#enable-keptn-for-your-cluster)

Reviewer: Setting the annotation for the namespace is not required if Keptn is used only for executing `KeptnTasks`, so we can remove the `enable` step here

Author: Who knew?  It's fixed.

title: Keptn for Non-Kubernetes Applications
description: Using Keptn with Non-Kubernetes Applications
weight: 95
---

Keptn Tasks can be triggered for workloads and applications
that are deployed outside of Kubernetes.
For example, Keptn could trigger load and performance tests
for an application that is deployed on a virtual machine.

To do this:

1. [Install and enable a Kubernetes cluster](#install-and-enable-a-kubernetes-cluster)
1. [Create a KeptnTaskDefinition](#create-a-keptntaskdefinition)
1. [Create and apply a KeptnTask](#create-and-apply-a-keptntask)
1. [Re-run the KeptnTask](#re-run-the-keptntask)

## Install and enable a Kubernetes cluster

You must still
[install](../install/install.md/#use-helm-chart)
and
[enable](../install/install.md/#enable-keptn-for-your-cluster)
Keptn on a Kubernetes cluster,

> but this can be a very lightweight, single-node KinD cluster; see

Reviewer 2: I really don't think we need instructions on how to create / run a kubernetes cluster. If a potentially user does not / cannot have a running cluster, they should first start learning what kubernetes is and how to install it (many many options). We can assume that Keptn users are technically savvy enough to be past that point.

That said, I do think pointing to an "as close to official small kubernetes distribution like kind" is useful.

Author: All I did here was provide links to the existing documentation about install and enable (the latter is Keptn specific, after all) and, just below, point to the instructions we have for setting up KinD.  If the user knows how to do these things, they don't need to click on those links.  Everyone agrees that we need to provide installation instructions in the Getting Started Guide, even though it is duplicating information that is elsewhere so requires double-maintenance.  And, if I'm new to Keptn, it's not totally obvious that I need a k8s cluster running Keptn even if all I want to do is run some tasks before and after my deployment that is not on k8s.  It seems to me that this just sets the stage, as it were.  We'll discuss ;-)

[Create local Kubernetes cluster](../install/k8s.md/#create-local-kubernetes-cluster).
Keptn only triggers on-demand `KeptnTask` resources
so resource utilization is minimal.

TODO: How is this cluster associated with the VM

> where the deployment is running?

Reviewer 2: I don't understand this comment. I think you are trying to ask "how does the VM know it is part of the cluster"? Well you installed Kind on your local machine so that's a single node cluster. Again, this is a detail which kubernetes abstracts from you. You ask it to run a Kubernetes job and hte job is scheduled to a suitable node. You don't (and probably shouldn't) care which node on the cluster. It is the Kubernetes scheduler's job to find you a suitable node.

Author: When you say "local machine", do you mean, say, my laptop or the VM/baremetal/whatever machine where the deployment runs?  If Kind is installed on the machine where the deployment runs, this makes sense but this doc doesn't explicitly say that.  We'll talk ;-)

Author: Removing this TODO for now.  We can revisit this later.

title: Keptn for Non-Kubernetes Applications
description: Using Keptn with Non-Kubernetes Applications
weight: 95
---

Keptn Tasks can be triggered for workloads and applications
that are deployed outside of Kubernetes.
For example, Keptn could trigger load and performance tests
for an application that is deployed on a virtual machine.

To do this:

1. [Install and enable a Kubernetes cluster](#install-and-enable-a-kubernetes-cluster)
1. [Create a KeptnTaskDefinition](#create-a-keptntaskdefinition)
1. [Create and apply a KeptnTask](#create-and-apply-a-keptntask)
1. [Re-run the KeptnTask](#re-run-the-keptntask)

## Install and enable a Kubernetes cluster

You must still
[install](../install/install.md/#use-helm-chart)
and
[enable](../install/install.md/#enable-keptn-for-your-cluster)
Keptn on a Kubernetes cluster,
but this can be a very lightweight, single-node KinD cluster; see
[Create local Kubernetes cluster](../install/k8s.md/#create-local-kubernetes-cluster).
Keptn only triggers on-demand `KeptnTask` resources
so resource utilization is minimal.

> TODO: How is this cluster associated with the VM

Reviewer 3: ```suggestion

```
I would remove this for now

Author: I can do that -- would love to get this much merged. actually.  But are we concerned that the instructions for doing this may be incomplete?

Reviewer 3: I would avoid trying to explain more because the "remote" setups could differ so much that it does not make sense for us to even try... we give you a place where to run anything you like (aka a container), so you can use it to connect and run stuff on anything you want as long as you know how to do it...

> where the deployment is running?

Reviewer 3: ```suggestion

```

Author: This comment makes no sense -- deleting the second half of the question in the TODO block?  Please clarify.

      - 'hello world'
```

This example uses the `container-runtime` runner,
but you can instead use the `deno-runtime` or `python-runtime` runner.
See
[Runners and containers](tasks/#runners-and-containers)
for more information.

## Create and apply a KeptnTask

You must manually create the
[KeptnTask](../yaml-crd-ref/task.md) resource.
In the standard operating mode, when Keptn is managing workloads,
the creation of the `KeptnTask` resource is automatic.

> Moreover, each time you want to execute a `KeptnTask` resource,

Reviewer 4: ```suggestion
Moreover, each time you want to execute a `KeptnTask`,
```

> you must manually create a a new (and uniquely named) `KeptnTask` resource.

Author: done

      - 'hello world'
```

This example uses the `container-runtime` runner,

> but you can instead use the `deno-runtime` or `python-runtime` runner.

Reviewer 4: can we add links to the descriptions of container-runtime, deno-runtime and python-runtime runners? So the users can click on it and read what the runners actually do

Author: done

## Create and apply a KeptnTask

You must manually create the
[KeptnTask](../yaml-crd-ref/task.md) resource.
In the standard operating mode, when Keptn is managing workloads,
the creation of the `KeptnTask` resource is automatic.

Moreover, each time you want to execute a `KeptnTask` resource,

> you must manually create a a new (and uniquely named) YAML file

Reviewer 4: ```suggestion
you must manually create a a new (and uniquely named) CR
```

> you must manually create a a new (and uniquely named) YAML file
> that describes that resource.

Reviewer 4: ```suggestion
you must manually create a a new (and uniquely named) `KeptnTask` resource.
```

      - 'hello world'
```

This example uses the `container-runtime` runner
because it allows the most flexibility
but you can also use the `deno-runtime` or `python-runtime` runners.

## Create and apply a KeptnTask

You must manually create the
[KeptnTask](../yaml-crd-ref/task.md)
the resource.
In the standard operating mode, when Keptn is managing workloads,
the creation of the `KeptnTask` resource is automatic.
Here though, we must create it ourselves.

Moreover, you must create a new (and uniquely named)
`KeptnTask` resource
each time you want to rerun this task
Each time you want to execute a `KeptnTask` resource,
you must manually create a
a new (and uniquely named)
[KeptnTask](../yaml-crd-ref/task.md)

> YAML file to describe that resource.

Reviewer 2: This is wrong. It's not the YAML filename on disk that matters but the `metadata.name` field which must be unique. `kubectl apply` applies the _contents_ of the YAML file. Kubernetes never knows about the filename

Author: This was corrected in the process of updating the syntax of this CRD

    workloadVersion: "1.0.0"
```

You can then apply this YAML file with the following command:

> ```yaml

Reviewer 4: ```suggestion
```shell
```

> kubectl --apply test-task.yaml

Reviewer 4: ```suggestion
kubectl apply -f test-task.yaml -n my-keptn-annotated-namespace
```

```

Applying this file causes Keptn to create a Job and a Pod
and run the executables defined
in the associated `KeptnTaskDefinition` resource.

Use the following commands to show the current status of the jobs:

```shell

> kubectl get keptntasks 
> kubectl get pods

Reviewer 4: ```suggestion
kubectl get keptntasks -n my-keptn-annotated-namespace
kubectl get pods -n my-keptn-annotated-namespace
```

```

## Re-run the KeptnTask

For subsequent KeptnTask runs,
the `KeptnTask` name and version fields must be unique,
so copy the `KeptnTask` yaml file you have and update the

> `metadata.name` field:

Reviewer 4: ```suggestion
`metadata.name` field.
```

A standard practice is to just increment the value of the suffix field.
For example, you could create a `test-task-2.yaml` file
with the `metadata.name` field set to `runhelloworld2`:

> ```yaml

Reviewer 3: I know this is not modified here , but should we profit from the fact that we touch this doc to embed the yaml instead? 
We could edit [this sample](https://github.com/keptn/lifecycle-toolkit/blob/main/lifecycle-operator/config/samples/lifecycle_v1alpha3_keptntask.yaml) and point to it

Author: I think we should merge this PR -- it's more than a month old -- then do another PR to make this change.  @mowies complains about the tendency to do one more thing while the file is open...

    workloadVersion: "1.0.0"
```

TODO: This file does not match what I see in the API Reference.
See specific comments in the `KeptnTask` reference page.
When we resolve those issues, I will modify this file appropriately.

You can then apply this YAML file with the following command:

```yaml
kubectl --apply test-task.yaml
```

Applying this file causes Keptn to create a Job and a Pod
and run the executables defined
in the associated `KeptnTaskDefinition` resource.

Use the following commands to show the current status of the jobs:

```shell
kubectl get keptntasks 
kubectl get pods
```

## Re-run the KeptnTask

For subsequent KeptnTask runs,
the `KeptnTask` name and version fields must be unique,
so copy the `KeptnTask` file you have and update the following fields:

- `name`
- `spec.appVersion`

> - `spec.workloadVersion`

Reviewer 2: There was a mismatch with the v1alpha3 spec which has now been resolved (see https://github.com/keptn/lifecycle-toolkit/issues/2120). The `spec.*` fields are removed now, defining the app & workload details only inside `spec.context` will work for v1alpha3

Author: Okay, this is new information.  I was basically just moving the existing page from the "Tutorials" section to the "User Guides" section so carrying through the information that was there.  I can fix this.

Author: This content has now been reconciled with the revised v1alpha3 spec

    workloadName: "my-workload"
    workloadVersion: "1.0.1"
```

You can then apply this file with the following command:

> ```yaml

Reviewer 4: ```suggestion
```shell
```

> kubectl --apply test-task-2.yaml

Reviewer 4: ```suggestion
kubectl apply -f test-task-2.yaml -n my-keptn-annotated-namespace
```
